include/pybind11/detail/internals.h

```diff
@@ -622,26 +709,16 @@ class internals_pp_manager {
         : holder_id_(id), on_fetch_(on_fetch) {}
 
     std::unique_ptr<InternalsType> *get_or_create_pp_in_state_dict() {
-        error_scope err_scope;
-        dict state_dict = get_python_state_dict();
-        auto internals_obj
-            = reinterpret_steal<object>(dict_getitemstringref(state_dict.ptr(), holder_id_));
-        std::unique_ptr<InternalsType> *pp = nullptr;
-        if (internals_obj) {
-            void *raw_ptr = PyCapsule_GetPointer(internals_obj.ptr(), /*name=*/nullptr);
-            if (!raw_ptr) {
-                raise_from(PyExc_SystemError,
-                           "pybind11::detail::internals_pp_manager::get_pp_from_dict() FAILED");
-                throw error_already_set();
-            }
-            pp = reinterpret_cast<std::unique_ptr<InternalsType> *>(raw_ptr);
-            if (on_fetch_ && pp) {
-                on_fetch_(pp->get());
-            }
-        } else {
-            pp = new std::unique_ptr<InternalsType>;
-            // NOLINTNEXTLINE(bugprone-casting-through-void)
-            state_dict[holder_id_] = capsule(reinterpret_cast<void *>(pp));
+        // The `unique_ptr<InternalsType>` output is leaked on interpreter shutdown. Once an
+        // instance is created, it will never be deleted until the process exits (compare to
+        // interpreter shutdown in multiple-interpreter scenarios).
+        // Because we cannot guarantee the order of destruction of capsules in the interpreter
+        // state dict, leaking avoids potential use-after-free issues during interpreter shutdown.
+        auto *pp
+            = atomic_get_or_create_in_state_dict<std::unique_ptr<InternalsType>,
+                                                 /*LeakOnInterpreterShutdown=*/true>(holder_id_);
+        if (on_fetch_ && pp) {
+            on_fetch_(pp->get());
```

Comment:
For internals, this used to only run when the pointer already existed within the state dict.  It did not run for newly created internals.

Reply:
Thanks for catching this! I spent a while carefully looking at the `atomic_get_or_create_in_state_dict` function, then created commit 1b6a857aa5d92f1239754b3bcc7832aa4e0c5994 to fix this issue, and apply some more minor changes. See commit message.

I still need to look at the latest versions of the new tests.